### ISSUE TITLE
the performance is bad...

### ISSUE THREAD

**Issue by coollofty** [radioflash/ljarray#1]
jit.off()

local s = os.clock()
print(s)
for i=1,100000 do
    a:insert(1)
    a:insert(2)
    a:remove(0)
    a:remove(0)
end
print('elapse', os.clock() - s)

--elapse  0.629
------------------compare--------------------
a = {}
local s = os.clock()
print(s)
for i=1,100000 do
    table.insert(a, 1)
    table.insert(a, 2)
    table.remove(a, 1)
    table.remove(a, 1)
end
print('elapse', os.clock() - s)

--elapse  0.02

if the jit is enabled, the using time is 0.003 for ljarray and 0.0019 for lua table

**Comment by radioflash** [comment]
First, thanks for your feedback.

If you can tell me more about your use-case and what kind of data you expect I might be able to give you more specific advice.

I have looked into this, and the core of this issue is that:
- LuaJIT FFI relies on JIT compilation for good performance
  Accessing FFI values ("cdata" in LuaJIT parlance) is very fast for JIT compiled code-- it is able to read and write cdata-variables basically as fast as C-code. In interpreted mode, on the other hand, there is quite a bit of marshaling involved (for every cdata read/write).
- The overhead of the accessor functions for ljarray code (which grow the array on insertions if necessary and/or move the data around) is negligible when JIT compiled, but quite visible when interpreted. This is further exacerbated by the fact that the ljarray itself is an FFI-value (which allows nesting), and every access off its properties (size and capacity) is much more expensive in interpreted mode.

There are ways to improve the performance even in interpreted mode, but those are mostly tradeoffs-- you could, for example, bypass the ljarray accessor functions and write to it directly ("a.data[i] = value;"). But this robs you of bounds-checking and automatic resizing and is still slower than using a Lua table in **interpreted** mode.

TL;DR: If you know that your code will be interpreted only, and your values are not passed to C-functions (or you have another reasons to prefer FFI-data), you should probably just use native Lua tables/types.

Again, if you could tell me what kind of data you expect/how you access it, I might be able to give you more specific advice.

Some hard data on FFI-performance (not involving ljarray, the FFI-Array is just the equivalent to a C-array):
$ luajit uncompiled_ffi.txt
JIT ON:
        sum with FFI-Array:     0.009
        sum with Lua-Table:     0.119
JIT OFF:
        sum with FFI-Array:     2.441
        sum with Lua-Table:     0.283

[uncompiled_ffi.txt](https://github.com/radioflash/ljarray/files/43686/uncompiled_ffi.txt)

**Comment by coollofty** [comment]
Thanks for fast reply. really I have not a use-case before I met ljarray, it is a very novel usage for FFI library. So, I made a simple test, it is only a very simple test, no purpose :) . But, I am particularly concerned about the performance of the JIT closed, because in iOS, the JIT also closed. That's all
